usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -63,7 +65,10 @@ const char label[] = "IBM ICA  PKCS #11";
 static pthread_mutex_t rngmtx = PTHREAD_MUTEX_INITIALIZER;
 
 #define LIBICA_SHARED_LIB "libica.so.3"
-#define BIND(dso, sym)  (*(void **)(&p_##sym) = dlsym(dso, #sym))
+#define BIND(dso, sym)  do {                                             \
+                            if (p_##sym == NULL)                         \
```

Comment:
wrt multi intance token, i think all global token data should either be moved to the token data structure or locked. Those kind of checks still leave a small window for race.

Reply:
Well, I thought about this, but its not really needed.
Token initialization is done during C_Initialize. The tokens are initialized one after the other (for loop that calls DL_Load_and_Init() ). The whole initialization process is locked using the GlobMutex in api_interface.c, so it can't run concurrently at the same time.
Thus it can not happen that two tokens get the symbols from the libraries at the same time. If 2 instances of the same token are initialized one after the other, the second one will either see that the symbol has already been loaded (ICA token), or will load the same symbol again (EP11), which does not hurt either.
So I would tend to leave it as is.

Reply:
ah okay. i assumed the token_specific_init function was actually called at CTokenInit, but your right. C_Initialize should be locked anyway.